### PR TITLE
Refactor ChainMetadata types

### PR DESCRIPTION
**Pull request by jmrossy**
### Description
- Refactor ChainMetadata types into new folder
- Create metadata extensions for deployment artifacts, and agent config
- Remove explicit ChainMetadata interface and infer from zod schema instead

### Related issues

Implements new schema discussed in:
- https://github.com/hyperlane-xyz/hyperlane-monorepo/issues/1867
- https://github.com/hyperlane-xyz/hyperlane-monorepo/pull/2203
- https://github.com/hyperlane-xyz/hyperlane-monorepo/issues/2214

Fixes https://github.com/hyperlane-xyz/hyperlane-monorepo/issues/2214

### Backward compatibility

**No**: users of chain metadata configs would need to slightly modify their shape (the `rpcUrl` field)

### CODE REVIEW

File: typescript/sdk/src/consts/environments/index.ts [hunk 1]
```diff
@@ -17,4 +18,4 @@ export type HyperlaneEnvironmentChain<E extends HyperlaneEnvironment> = Extract<
 export const hyperlaneContractAddresses = objMerge(
   hyperlaneEnvironments.testnet,
   hyperlaneEnvironments.mainnet,
-);
+) as Record<CoreChainName, typeof hyperlaneEnvironments['mainnet']['ethereum']>;
```
Comment (jmrossy) [inline typescript/sdk/src/consts/environments/index.ts:21]: Just a drive-by change to fix the type on this object

File: typescript/sdk/src/metadata/userInterfaceExtension.ts [hunk 2]
```diff
@@ -0,0 +1,35 @@
+import { z } from 'zod';
+
+import { ChainMetadataSchema } from './chainMetadataTypes';
+
+/**
+ * Extension of ChainMetadataSchema with additional properties for user interfaces
+ */
+export const UserInterfaceMetadataExtSchema = z.object({
+  displayName: z
```
[the remark this answers is not in the document]
Comment (jmrossy) [reply typescript/sdk/src/metadata/userInterfaceExtension.ts:9]: We can yes and I'm fine with that. Breaking them out was in response to a comment @mattiecnvr left [here](https://github.com/hyperlane-xyz/hyperlane-monorepo/issues/1867#issuecomment-1524035307) about how `displayName` is unused in the agents. I would have broken out things like `gasCurrencyCoinGeckoId` too but they're used in the SDK itself. What would you prefer?

File: typescript/sdk/src/metadata/chainMetadataTypes.ts [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+  protocol: z
+    .nativeEnum(ProtocolType)
+    .describe(
+      'The type of protocol used by this chain. See ProtocolType for valid values.',
+    ),
+  chainId: z
+    .number()
+    .positive()
+    .describe(`The chainId of the chain. Uses EIP-155 for EVM chains`),
+  domainId: z
```
[the remark this answers is not in the document]
Comment (jmrossy) [reply typescript/sdk/src/metadata/chainMetadataTypes.ts:38]: Yeah good call-out. The were optional but as I was implementing multi-env support in the warp UI and finding that some chains don't have a definitive `chainID`, it changed my mind on this. I'd recommend we be more strict about explicitly setting these. WDYT?

[the remark this answers is not in the document]
Comment (nambrot) [reply typescript/sdk/src/metadata/chainMetadataTypes.ts:38]: Don't feel strongly, i believe this will also apply to cosmos chains. On the margin, would probably prefer to keep it optional for PI EVM folks

File: typescript/sdk/src/metadata/userInterfaceExtension.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (nambrot) [reply typescript/sdk/src/metadata/userInterfaceExtension.ts:9]: I'd keep them in `ChainMetadata` for simplicity rn

File: typescript/sdk/src/metadata/chainMetadataTypes.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (jmrossy) [reply typescript/sdk/src/metadata/chainMetadataTypes.ts:38]: Okay, I'll revert back to optional, np

File: typescript/sdk/src/metadata/userInterfaceExtension.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (jmrossy) [reply typescript/sdk/src/metadata/userInterfaceExtension.ts:9]: Okay will do

File: typescript/sdk/src/metadata/chainMetadataTypes.ts [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+    .number()
+    .positive()
+    .describe(
+      'The domainId of the chain, should generally default to `chainId`. Consumer of `ChainMetadata` should use this value if present, but otherwise fallback to `chainId`.',
+    ),
+  name: z
+    .string()
+    .describe(
```
[the remark this answers is not in the document]
Comment (jmrossy) [reply typescript/sdk/src/metadata/chainMetadataTypes.ts:48]: Done.

File: typescript/sdk/src/metadata/agentConfig.ts [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+    };
+    configs[chain] = config;
+  }
+  return configs;
+}
+
+/**
+ * Deprecated agent config shapes.
+ * See https://github.com/hyperlane-xyz/hyperlane-monorepo/issues/2215
+ */
```
Comment (jmrossy) [inline typescript/sdk/src/metadata/agentConfig.ts:84]: All code below this is unchanged, copied from typescript/sdk/src/agents/types.ts

**Review by asaj — APPROVED** [review]

File: typescript/sdk/src/consts/environments/index.ts (hunk 1 above)
Comment (jmrossy) [reply typescript/sdk/src/consts/environments/index.ts:21]: objMerge is dropping the type

File: typescript/sdk/src/consts/chainMetadata.ts [hunk 6]
```diff
@@ -781,6 +607,86 @@ export const test3: ChainMetadata = {
...
+    },
+  ],
+  blocks: {
+    confirmations: 1,
+    reorgPeriod: 0,
+    estimateBlockTime: 0.4,
+  },
+};
+
+export const solanatestnet: ChainMetadata = {
```
[the remark this answers is not in the document]
Comment (jmrossy) [reply typescript/sdk/src/consts/chainMetadata.ts:638]: Will fix!
